### DISCUSSION TITLE
Enable RLS by default on all new tables

### DISCUSSION THREAD

**Discussion by licarth** [supabase/supabase#21747] [Questions] [answered]
Row-level security is by default disabled on new Postgres tables.

#9690 only works when tables are created from the UI.

I'm using Prisma to manage my database schema (and I guess I'm not the only one to use an ORM), and in this case RLS is disabled on new tables.

I think RLS should be enabled by default to avoid leaking data publicly, by accident. It's what Firebase is doing with their Firestore security rules.

To me, it's something that would deter us from using Supabase Postgres as we scale up to more employees. I'm currently manually doing this, but it's easy to forget.

**Comment by GaryAustin1** [comment, marked as answer]
When you use Prisma it is creating the tables to Postgres directly with SQL.  Supabase is not involved.  Seems like this would be Prisma thing to solve.  Other wise Postgres itself would have to change and good luck making that happen.

Going to move this to a discussion as it does not seem to be something Supabase could address as their code is not between Prisma and Postgres.

**Comment by kiwicopple** [comment]
NOTE: this is answered here: https://github.com/orgs/supabase/discussions/21747#discussioncomment-8668775

To keep the Feature Request section clean, I will migrate this to a General discussion 👍

**Comment by licarth** [comment]
@GaryAustin1 Ok, my bad, I did not understand that row-level security was [existing in PostgreSQL itself](https://www.postgresql.org/docs/current/ddl-rowsecurity.html). I thought this was some feature that Supabase was providing on top of it.

It's possible to set it via some SQL statements `alter table ... enable row level security`. But indeed Prisma [does not seem to support](https://github.com/prisma/prisma/issues/12735) this yet.

So I'll do that when creating the tables with Prisma. Will probably require some manual SQL migrations, but at least I can try to automate them and they'll be replicated accross environments running the same SQL migrations.

**Comment by GraemeFulton** [comment]
Here's a sql i used to activate RLS on loads of tables at once:
```
-- Enable row-level security (RLS) dynamically on all base tables in the public schema
DO $$
DECLARE
  table_record RECORD;
BEGIN
  FOR table_record IN 
    SELECT table_name 
    FROM information_schema.tables 
    WHERE table_schema = 'public' 
      AND table_type = 'BASE TABLE'
  LOOP
    EXECUTE format('ALTER TABLE public.%I ENABLE ROW LEVEL SECURITY;', table_record.table_name);
  END LOOP;
END $$;
```
wrote more in [this note](https://prototypr.io/note/activate-rls-supabase-all)

**Comment by phgn0** [comment]
I personally find it ABSOLUTELY CRAZY that Supabase tables are exposed by default if you create them through an ORM like Prisma or Drizzle.

People (like me) will forget to run the manual SQL command to enable RLS after some time. This is not a solution for a platform that's supposed to make development easy and have security by default (which it does not, evidently).

**Reply by GaryAustin1** [reply]
When you create a table with Supabase it does enable RLS by default...
<img width="696" height="567" alt="image" src="https://github.com/user-attachments/assets/555dc489-df79-49d1-8ae2-95b740f6e29d" />

But you are using the Postgres behind the scenes with your ORM.  Supabase is not involved in creating the tables as you are talking directly to the Postgres database.

**Reply by GaryAustin1** [reply]
Also there is an option when you set up Supabase to turn of the REST API totally if you are only planning to use direct database connections.

**Reply by tyler-lemke** [reply]
I am also baffled by this. This is a supabase problem in that it exposes the REST API itself. 

You have to go in and set this up with the following:
 Dashboard > Project > Authentication > Auto-enable RLS for new tables > Learn More > Click "create ensure_rls trigger"
 
That createensure_rls trigger should be default enabled, not opt-in if your database has the REST API enabled.  Am I missing something here? I'm newer to Postgres but a security posture like this in 2026 seems meh at best.

**Reply by GaryAustin1** [reply]
> I am also baffled by this. This is a supabase problem in that it exposes the REST API itself.
> 
> You have to go in and set this up with the following: Dashboard > Project > Authentication > Auto-enable RLS for new tables > Learn More > Click "create ensure_rls trigger"
> 
> That createensure_rls trigger should be default enabled, not opt-in if your database has the REST API enabled. Am I missing something here? I'm newer to Postgres but a security posture like this in 2026 seems meh at best.

Probably relevant... https://github.com/orgs/supabase/discussions/45329
